language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -129,436 +110,178 @@ impl AbstractState {
         self.locals.insert(idx, value);
     }
 
-    /// Return true if self may safely be destroyed
-    pub fn is_safe_to_destroy(&self) -> bool {
-        self.locals.values().all(|x| x.is_safe_to_destroy())
-            && self.globals.values().all(|x| x.is_empty())
+    /// checks if a local may be safely destroyed
+    pub fn is_local_safe_to_destroy(&self, idx: LocalIndex) -> bool {
+        match self.locals[&idx] {
+            AbstractValue::Reference(_) => true,
+            AbstractValue::Value(Kind::All) | AbstractValue::Value(Kind::Resource) => false,
+            AbstractValue::Value(Kind::Unrestricted) => !self.is_local_borrowed(idx),
+        }
     }
 
-    /// Return true if local@idx may safely be destroyed
-    pub fn is_local_safe_to_destroy(&self, idx: LocalIndex) -> bool {
-        self.local(idx).is_safe_to_destroy()
+    /// checks if the stack frame of the function being analyzed can be safely destroyed
+    pub fn is_frame_safe_to_destroy(&self) -> bool {
+        self.locals.values().all(|x| x.is_unrestricted_value())
+            && !self.is_nonce_borrowed(self.frame_root)
```

Comment:
This code is not immediately obvious to me.

Reply:
I have expanded the comment explaining what the function does.